File: gnu/gcc/libgomp/testsuite/libgomp.fortran/appendix-a/a.16.1.f90
! { dg-do run }
        REAL FUNCTION WORK1(I)
            INTEGER I
            WORK1 = 1.0 * I
            RETURN
        END FUNCTION WORK1

        REAL FUNCTION WORK2(I)
            INTEGER I
            WORK2 = 2.0 * I
            RETURN
        END FUNCTION WORK2

        SUBROUTINE SUBA16(X, Y, INDEX, N)
        REAL X(*), Y(*)
        INTEGER INDEX(*), N
        INTEGER I
!$OMP PARALLEL DO SHARED(X, Y, INDEX, N)
          DO I=1,N
!$OMP ATOMIC
              X(INDEX(I)) = X(INDEX(I)) + WORK1(I)
            Y(I) = Y(I) + WORK2(I)
          ENDDO
      END SUBROUTINE SUBA16

      PROGRAM A16
        REAL X(1000), Y(10000)
        INTEGER INDEX(10000)
        INTEGER I
        DO I=1,10000
          INDEX(I) = MOD(I, 1000) + 1
          Y(I) = 0.0
        ENDDO
        DO I = 1,1000
          X(I) = 0.0
        ENDDO
        CALL SUBA16(X, Y, INDEX, 10000)
        DO I = 1,10
          PRINT *, "X(", I, ") = ", X(I), ", Y(", I, ") = ", Y(I)
        ENDDO
      END PROGRAM A16

File: gnu/gcc/libgomp/testsuite/libgomp.c/appendix-a/a.16.1.c
/* { dg-do run } */

#include <stdio.h>

float
work1 (int i)
{
  return 1.0 * i;
}

float
work2 (int i)
{
  return 2.0 * i;
}

void
a16 (float *x, float *y, int *index, int n)
{
  int i;
#pragma omp parallel for shared(x, y, index, n)
  for (i = 0; i < n; i++)
    {
#pragma omp atomic
      x[index[i]] += work1 (i);
      y[i] += work2 (i);
    }
}
int
main ()
{
  float x[1000];
  float y[10000];
  int index[10000];
  int i;
  for (i = 0; i < 10000; i++)
    {
      index[i] = i % 1000;
      y[i] = 0.0;
    }
  for (i = 0; i < 1000; i++)
    x[i] = 0.0;
  a16 (x, y, index, 10000);
  for (i = 0; i < 10; i++)
    printf ("x[%d] = %f, y[%d] = %f\n", i, x[i], i, y[i]);
  return 0;
}
